### PR TITLE
Use static configurations for building a vertical

### PR DESCRIPTION
**Pull request by chcosta**
Disables the dynamic determination of groups and uses the static configuration files produced by src\Tools.

/cc @weshaggard @ericstj

### CODE REVIEW

**Comment by chcosta** [comment]
Looks like my changes may have affected the package build.  Investigating

**Comment by chcosta** [comment]
Looking into this issue - 

17:34:48 EXEC : error : The process cannot access the file '/mnt/resource/j/workspace/dotnet_corefx/dev_eng/centos7.1_debug_prtest/pkg/baseline/project.lock.json' because it is being used by another process. [/mnt/resource/j/workspace/dotnet_corefx/dev_eng/centos7.1_debug_prtest/pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]
17:34:48 EXEC : error : The process cannot access the file '/mnt/resource/j/workspace/dotnet_corefx/dev_eng/centos7.1_debug_prtest/pkg/baseline/project.lock.json' because it is being used by another process. [/mnt/resource/j/workspace/dotnet_corefx/dev_eng/centos7.1_debug_prtest/pkg/Microsoft.NETCore.Platforms/Microsoft.NETCore.Platforms.pkgproj]

**Comment by ericstj** [comment]
You can delete the following from baseline.props:
https://github.com/dotnet/corefx/blob/cd17e98cc1b521d97047893dd7d14344c250341c/pkg/baseline/baseline.props#L9-L12

That's causing restore to happen during build.

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj [hunk 1]
```diff
@@ -88,6 +88,8 @@
 
   <ItemGroup>
     <IgnoredReference Include="System.Private.CoreLib" />
+    <IgnoredReference Include="System.Threading.Overlapped" />
```
Comment (ericstj) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]: Why?

File: dir.traversal.targets [hunk 2]
```diff
@@ -28,33 +28,29 @@
...
   <!-- build vertical requires the FindBestConfiguration task which is currently in CoreFx.Tools.dll -->
-  <Import Project="buildvertical.targets" Condition="Exists('$(CoreFxToolsTaskDir)CoreFx.Tools.dll')" />
+<!-->  <Import Project="buildvertical.targets" Condition="Exists('$(CoreFxToolsTaskDir)CoreFx.Tools.dll')" /> -->
 
   <Target Name="BuildAllProjects">
     <PropertyGroup>
       <DefaultBuildAllTarget Condition="'$(DefaultBuildAllTarget)'==''">$(MSBuildProjectDefaultTargets)</DefaultBuildAllTarget>
     </PropertyGroup>
 
-    <MSBuild Targets="AnnotateProjectsWithConfiguration"
```
Comment (ericstj) [inline dir.traversal.targets]: Why are you turning this off?

File: Tools-Override/resolveContract.targets [hunk 3]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (weshaggard) [inline Tools-Override/resolveContract.targets]: Can you please put an unmodified version of this in an earlier commit so we can see what changed?

File: dir.props [hunk 4]
```diff
@@ -16,13 +16,15 @@
     <!-- Temp change to make OS X build behave as a Linux build -->
     <OsEnvironment Condition="'$(OsEnvironment)'=='' AND '$(OS)'=='OSX'">Unix</OsEnvironment>
     <OsEnvironment Condition="'$(OsEnvironment)'==''">$(OS)</OsEnvironment>
+    <OriginalConfiguration>$(Configuration)</OriginalConfiguration>
```
Comment (weshaggard) [inline dir.props]: Where is this used and why do we need it?

File: dir.props [hunk 5]
```diff
@@ -197,6 +186,8 @@
     </When>
   </Choose>
 
+  <Import Project="$(BinDir)tools/configuration/configuration.props" Condition="Exists('$(BinDir)tools/configuration/configuration.props')" />
```
Comment (weshaggard) [inline dir.props]: Doesn't this need to be above the Choose statement looking at ConfigurationGroup? Otherwise that will never be set.

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj (hunk 1 above)
Comment (weshaggard) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]: Those projects don't build yet so I asked Chris to just ignore them for now until we get them building and then we can remove these.

File: src/Microsoft.CSharp/src/Configurations.props [hunk 6]
```diff
@@ -1,8 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <PropertyGroup>
-    <BuildConfigurations>
-      netstandard1.7;
-    </BuildConfigurations>
+    <BuildConfigurations>netstandard1.7;</BuildConfigurations>
```
Comment (weshaggard) [inline src/Microsoft.CSharp/src/Configurations.props]: Any reason you changed the formatting here?

File: src/src.builds [hunk 7]
```diff
@@ -1,8 +1,11 @@
 <?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
   <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
-  <ItemGroup Condition="'$(OS)' != 'Windows_NT'">
-    <ProjectExclusions Include="$(MSBuildThisFileDirectory)System.Memory/src/System.Memory.csproj" />
+  <ItemGroup>
+    <ProjectExclusions Condition="'$(OS)' != 'Windows_NT'" Include="$(MSBuildThisFileDirectory)**\src\System.Memory.csproj" />
+    <ProjectExclusions Include="$(MSBuildThisFileDirectory)**\src\System.Threading.Overlapped.csproj" />
```
Comment (weshaggard) [inline src/src.builds]: Lets make sure we have an issue to track fixing these projects and removing them from this list and the package ignore list.

File: buildvertical.targets.save [hunk 8]
```diff
@@ -28,7 +28,8 @@
           Outputs="fake"
           Returns="@(Project)"
           DependsOnTargets="TrimUnconvertedProjects"
```
Comment (weshaggard) [inline buildvertical.targets.save]: Is the plan to delete this file as you renamed it to ".save".

**Comment by weshaggard** [comment]
>It's unclear to me why you are doing this. It looks like you're undoing work that you already had working. Can you explain what was broken and how this fixes it?

Previously his change wasn't using the static configuration files at all and so local project builds didn't work and he also wasn't passing the right configuration values so while the basic full stack build was working for projects that had configurations there were a number of other scenarios not working.

File: dir.targets [hunk 9]
```diff
@@ -85,4 +74,10 @@
     </ItemGroup>
   </Target>
 
+  <Target Name="ReportGroups"
+          BeforeTargets="PrepareForBuild"
```
Comment (weshaggard) [inline dir.targets]: Why PrepareForBuild? I'm not sure that if that is right or wrong just asking why you picked that one?

File: buildvertical.targets.save (hunk 8 above)
Comment (chcosta) [inline buildvertical.targets.save]: I plan to delete this once we determine that we don't need any dynamic configuring.

File: dir.props (hunk 4 above)
Comment (chcosta) [inline dir.props]: no longer used

File: pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj (hunk 1 above)
Comment (ericstj) [inline pkg/Microsoft.Private.CoreFx.NETCoreApp/Microsoft.Private.CoreFx.NETCoreApp.pkgproj]: I see, let's file an issue to remove them.  This will impact testing.

private.corelib is intentionally missing, these two are bugs.  I don't want to get in the habit of suppressing missing dependencies since the consequences downstream are harder for folks to diagnose.

File: dir.traversal.targets (hunk 2 above)
Comment (ericstj) [inline dir.traversal.targets]: I don't think we should turn the whole thing off just to fix packages.builds.  The projects it builds do their own filtering.  If you'd like me to get this fixed I can do that but don't sacrifice progress here.

Comment (chcosta) [inline dir.traversal.targets]: I think that we'll need some of this in place to pass CI on non-Windows platforms.  I'm looking into it now.

File: buildvertical.targets [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    <ItemGroup><VerticalTargets Remove="@(VerticalTargets)" /></ItemGroup>
+    <MSBuild Targets="FindBestConfiguration"
+             Projects="$(_VerticalProject)"
+             Condition="'$(_OmitReason)' == ''">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="VerticalTargets" />
+    </MSBuild>
+
+    <PropertyGroup>
+      <_SelectedOSGroup>%(VerticalTargets.OSGroup)</_SelectedOSGroup>
```
Comment (ericstj) [inline buildvertical.targets]: We shouldn't need to handle OSGroup and TargetGroup specially.  Just let the project tell you its configuration.

File: dir.props [hunk 11]
```diff
@@ -22,12 +22,8 @@
     <!-- Temporarily omit code analysis -->
     <SkipImportCodeAnalysisTargets>true</SkipImportCodeAnalysisTargets>
 
-    <BuildConfiguration_TargetGroup>netcoreapp1.1</BuildConfiguration_TargetGroup>
-    <BuildConfiguration_TargetGroup Condition="'$(TargetGroup)' != ''">$(TargetGroup)</BuildConfiguration_TargetGroup>
-    <BuildConfiguration_OSGroup>$(OS)</BuildConfiguration_OSGroup>
-    <BuildConfiguration_OSGroup Condition="'$(OSGroup)' != ''">$(OSGroup)</BuildConfiguration_OSGroup>
     <!-- Hacky BuildConfiguration -->
-    <BuildConfiguration Condition="'$(BuildConfiguration)' == ''">$(BuildConfiguration_TargetGroup)-$(BuildConfiguration_OSGroup)</BuildConfiguration>
+    <BuildConfiguration>netcoreapp1.1-$(OS)</BuildConfiguration>
```
Comment (ericstj) [inline dir.props]: Condition?

File: Tools-Override/FrameworkTargeting.targets [hunk 12]
```diff
@@ -126,28 +126,22 @@
...
-      </ProjectReference>
-      <ProjectReference>
-        <AdditionalProperties Condition="'%(ProjectReference.OSGroup)'!=''">OSGroup=%(ProjectReference.OSGroup);%(ProjectReference.AdditionalProperties)</AdditionalProperties>
-      </ProjectReference>
-      <ProjectReference>
-        <AdditionalProperties Condition="'$(ConfigurationGroup)' != ''">ConfigurationGroup=$(ConfigurationGroup);%(ProjectReference.AdditionalProperties)</AdditionalProperties>
+<!-->        <UndefineProperties Condition="'$(ConfigurationGroup)' != ''">%(ProjectReference.UndefineProperties);Configuration</UndefineProperties> -->
       </ProjectReference>
 
+     <!-- Configuration property shortcuts -->
```
Comment (weshaggard) [inline Tools-Override/FrameworkTargeting.targets]: I'm not convinced that we even need to keep these shortcuts any longer. @ericstj do you think they are interesting for any scenarios now?

File: buildvertical.targets [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+                           PropertyValues="@(PropertyValue)"
+                           BuildConfigurations="$(BuildConfigurations)"
+                           BuildConfiguration="$(BuildConfiguration)" 
+                           Condition="'$(BuildConfigurations)' != ''"
+                           ContinueOnError="true">
+      <Output TaskParameter="BestConfiguration" ItemName="VerticalTargets" />                          
+    </FindBestConfiguration>  
+
+    <ItemGroup Condition="'$(BuildConfigurations)' == ''">
+      <VerticalTargets Include="$(MSBuildProjectFullPath)">
```
Comment (ericstj) [inline buildvertical.targets]: If BuildConfigurations is empty then lets set use the current BuildConfiguration as BestConfiguration.

Comment (ericstj) [inline buildvertical.targets]: Also, the identity of this item should be a configuration string and should be used directly.

File: buildvertical.targets [hunk 14]
```diff
@@ -47,15 +47,18 @@
     <PropertyGroup>
       <_SelectedOSGroup>%(VerticalTargets.OSGroup)</_SelectedOSGroup>
       <_SelectedTargetGroup>%(VerticalTargets.TargetGroup)</_SelectedTargetGroup>
+      <_Configuration>$(_SelectedTargetGroup)</_Configuration>
```
Comment (weshaggard) [inline buildvertical.targets]: Where are these coming from?

File: Tools-Override/FrameworkTargeting.targets (hunk 12 above)
Comment (ericstj) [inline Tools-Override/FrameworkTargeting.targets]: No we don't need them.  In fact we should try to eradicate all uses of AdditionalProperties and UndefineProperties other than Configuration.

File: buildvertical.targets [hunk 15]
```diff
@@ -138,16 +142,34 @@
     <PropertyGroup>
       <_SelectedTargetGroup>%(VerticalTargets.TargetGroup)</_SelectedTargetGroup>
       <_SelectedOSGroup>%(VerticalTargets.OSGroup)</_SelectedOSGroup>
+      <_Configuration>$(_SelectedTargetGroup)</_Configuration>
+      <_Configuration Condition="'$(_SelectedOSGroup)' != '' and '$(_SelectedOSGroup)' != 'AnyOS'">$(_Configuration)-$(_SelectedOSGroup)</_Configuration>
+      <_AdditionalProperties Condition="'$(_SelectedTargetGroup)' != ''">TargetGroup=$(_SelectedTargetGroup)</_AdditionalProperties>
+      <_AdditionalProperties Condition="'$(_SelectedOSGroup)' != '' and '$(_SelectedOSGroup)' != 'AnyOS'">OSGroup=$(_SelectedOSGroup);$(_AdditionalProperties)</_AdditionalProperties>
+      <_AdditionalProperties Condition="'$(_Configuration)' != ''">Configuration=$(_Configuration);$(_AdditionalProperties)</_AdditionalProperties>
```
Comment (weshaggard) [inline buildvertical.targets]: I think we should try and make it such that the only property we pass through is the Configuration property (and perhaps the archgroup if it is needed later). I want to try and avoid needing to know about all these individual properties here. They are easily parsed on the project side from the Configuration.

File: buildvertical.targets [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+             Condition="'$(_OmitReason)' == ''">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="VerticalTargets" />
+    </MSBuild>
+
+    <PropertyGroup>
+      <_SelectedOSGroup>%(VerticalTargets.OSGroup)</_SelectedOSGroup>
+      <_SelectedTargetGroup>%(VerticalTargets.TargetGroup)</_SelectedTargetGroup>
+      <_Configuration>$(_SelectedTargetGroup)</_Configuration>
+      <_Configuration Condition="'$(_SelectedOSGroup)' != '' and '$(_SelectedOSGroup)' != 'AnyOS'">$(_Configuration)-$(_SelectedOSGroup)</_Configuration>
```
Comment (ericstj) [inline buildvertical.targets]: Don't reconstruct, just use what you get from FindBestConfiguration.  It's already a configuration string.

File: buildvertical.targets [hunk 17]
```diff
@@ -138,16 +142,34 @@
...
-          <OSGroup Condition="'%(ProjectReference.OSGroup)' == ''">$(_SelectedOSGroup)</OSGroup>
+        <ProjectReference Condition="'%(ProjectReference.Identity)' == '$(_ProjectReferenceToAnnotate)' and '$(BuildConfigurations)' != ''">
+          <TargetGroup>$(_SelectedTargetGroup)</TargetGroup>
+          <OSGroup>$(_SelectedOSGroup)</OSGroup>
+          <Configuration>$(_Configuration)</Configuration>
+          <AdditionalProperties>$(_AdditionalProperties);%(ProjectReference.AdditionalProperties)</AdditionalProperties>
         </ProjectReference>
+
+        <ProjectReference Condition="'%(ProjectReference.Identity)' == '$(_ProjectReferenceToAnnotate)' and ('$(_SelectedOSGroup)'  == '' or '$(_SelectedOSGroup)' == 'AnyOS')">
+          <UndefineProperties>OSGroup;%(ProjectReference.UndefineProperties)</UndefineProperties>
```
Comment (weshaggard) [inline buildvertical.targets]: We shouldn't need to worry about undefining OSGroup/TargetGroup/Configuration group because I'm thinking we will eventually forcibly override and ignore them and always set them from Configuration. However I'm fine keeping this for now if it unblock your work.

File: buildvertical.targets.save [hunk 18]
```diff
@@ -0,0 +1,187 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (weshaggard) [inline buildvertical.targets.save]: I assume this .save file should be deleted now.

File: config.json [hunk 19]
```diff
@@ -283,7 +277,7 @@
         "os": {
           "description": "Sets OSGroup to the OS name where the build is going to run or it is set to the value passed by the user.",
           "settings": {
-            "FakeOSGroup": "default"
+            "OSGroup": "default"
```
Comment (weshaggard) [inline config.json]: Is there a specific reason this was needed? I want to understand it because I was thinking of removing this.

File: dir.props [hunk 20]
```diff
@@ -21,8 +21,10 @@
   <PropertyGroup>
     <!-- Temporarily omit code analysis -->
     <SkipImportCodeAnalysisTargets>true</SkipImportCodeAnalysisTargets>
+
     <!-- Hacky BuildConfiguration -->
-    <BuildConfiguration Condition="'$(BuildConfiguration)' == ''">netcoreapp1.1-$(OSEnvironment)</BuildConfiguration>
+    <BuildConfiguration>netcoreapp1.1-$(OS)</BuildConfiguration>
```
Comment (weshaggard) [inline dir.props]: I think this property should be defined lower right above where we import the generated config props. Also here is what I'm thinking we should default it to. 

```
  <PropertyGroup>
    <ArchGroup Condition="'$(ArchGroup)'==''">x64</ArchGroup>
    <BuildConfiguration Condition="'$(BuildConfiguration)' == ''">netcoreapp-$(OSEnvironment)-Debug-$(ArchGroup)</BuildConfiguration>
  </PropertyGroup>
```
Although we could end up using $(OS) instead of $(OSEnviroment) if needed for some reason.

File: dir.props [hunk 21]
```diff
@@ -156,35 +158,29 @@
...
-    <When Condition="'$(ConfigurationGroup)'=='Debug'">
-      <PropertyGroup>
-        <DebugSymbols Condition="'$(DebugSymbols)' == ''">true</DebugSymbols>
-        <Optimize Condition="'$(Optimize)' == ''">false</Optimize>
-        <DebugType Condition="'$(DebugType)' == ''">full</DebugType>
-        <DefineConstants>$(DefineConstants),DEBUG,TRACE</DefineConstants>
+  <Import Project="$(BuildConfigurationImportFile)" Condition="Exists('$(BuildConfigurationImportFile)') and '$(BuildConfigurations)' != ''" /> 
+
+  <!-- Set up Default symbol and optimization for Configuration --> 
+  <Choose>
```
Comment (weshaggard) [inline dir.props]: Can you please file and issue to move these properties and the Targets* properties to the generated props files.

File: dir.targets [hunk 22]
```diff
@@ -13,25 +13,24 @@
...
+      <_UndefineProperties>%(ProjectReference.UndefineProperties)</_UndefineProperties>
+    </PropertyGroup>
     <ItemGroup>
-      <ProjectReference>
         <!--
         Always undefine TestTFM and FilterToTestTFM for all project reference as not needed for compilation and
         avoid bin clash tool to fail
         -->
-        <UndefineProperties>%(ProjectReference.UndefineProperties);TestTFM;FilterToTestTFM</UndefineProperties>
+      <ProjectReference Condition="'%(ProjectReference.Identity)' == '$(ProjectReference)'">
```
Comment (weshaggard) [inline dir.targets]: Where is the property `$(ProjectReference)` defined? I see `$(_ProjectReference)` above. Also what impact does this have on your changes I don't see the connection.

File: dir.targets [hunk 23]
```diff
@@ -85,4 +84,11 @@
     </ItemGroup>
   </Target>
 
+  <Target Name="ReportConfigurationErrorMessage"
+          BeforeTargets="ResolveProjectReferences"
+           Condition="'$(BuildConfigurations)' != ''" >
+    <Message Importance="High" Text="$(MSBuildProjectFullPath), Configuration: $(Configuration)" />
```
Comment (weshaggard) [inline dir.targets]: In my local changes I have also been dumping `$(BuildConfiguration)` and `$(BuildConfigurations)` to help with debugging it might be worth adding those and we can keep them until we get this all worked out and then we can turn them to Importance=low

File: dir.traversal.targets [hunk 24]
```diff
@@ -216,9 +215,43 @@
     <Error Condition="'$(MSBuildLastTaskResult)'=='false'" />
   </Target>
 
+ <Target Name="AnnotateProjectsWithConfiguration"
```
Comment (weshaggard) [inline dir.traversal.targets]: Is your assumption that this only happens for Project's which are themselve's traversals? Wouldn't we need this target for just plan projects as well?

File: src/System.Runtime.Extensions/ref/Configurations.props [hunk 25]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <BuildConfigurations>
```
Comment (weshaggard) [inline src/System.Runtime.Extensions/ref/Configurations.props]: Why did you need to add just the one configuration for this ref?

File: buildvertical.targets [hunk 26]
```diff
@@ -8,102 +8,92 @@
...
-    <ItemGroup>
-      <OmitProject Include="%(Project.Identity)" Condition="'%(Project.OmitReason)' != ''">
-        <OmitReason>%(Project.OmitReason)</OmitReason>
-      </OmitProject>
-      <Project Remove="@(OmitProject)" />
-    </ItemGroup>
+ <Target Name="AnnotateProjectsWithConfiguration" 
+         Returns="@(ProjectWithConfiguration)"> 
+    <PropertyGroup>
+      <_ProjectName>$(MSBuildProjectFullPath)</_ProjectName>
```
Comment (weshaggard) [inline buildvertical.targets]: Is there a reason you need this intermediate property?

File: buildvertical.targets [hunk 27]
```diff
@@ -8,102 +8,92 @@
...
+    </PropertyGroup>
+    <Message Importance="Low" Text="$(MSBuildProjectFullPath), Configuration - $(_BestConfiguration)" />
+    <ItemGroup> 
+       <ProjectWithConfiguration Include="$(_ProjectName)" Exclude="@(OmitProject)"> 
+         <AdditionalProperties Condition="'$(_BestConfiguration)' != '' and '$(_BestConfiguration)' != 'None'">Configuration=$(_BestConfiguration);%(ProjectWithConfiguration.AdditionalProperties)</AdditionalProperties>
+         <UndefineProperties Condition="'$(_BestConfiguration)' == '' or '$(_BestConfiguration)' == 'None'">Configuration;%(ProjectWithConfiguration.UndefineProperties)</UndefineProperties>
+       </ProjectWithConfiguration>
+       <ProjectWithConfiguration Remove="@(OmitProject)" />
```
Comment (weshaggard) [inline buildvertical.targets]: Why do you need to Remove OmitProject when you are excluding it above, or vice versa, I would expect one or the other.

File: buildvertical.targets [hunk 28]
```diff
@@ -8,102 +8,92 @@
...
-      <_VerticalProjectName>%(Project.FullPath)</_VerticalProjectName>
-      <_OmitReason>%(Project.OmitReason)</_OmitReason>
+      <_VerticalProject>$(MSBuildProjectFullPath)</_VerticalProject>
     </PropertyGroup>
 
-    <!-- Clear the VerticalTargets item -->
-    <ItemGroup><VerticalTargets Remove="@(VerticalTargets)" /></ItemGroup>
+    <!-- Clear the _Configuration item -->
+    <ItemGroup><_Configuration Remove="@(_Configuration)" /></ItemGroup>
     <MSBuild Targets="FindBestConfiguration"
```
Comment (weshaggard) [inline buildvertical.targets]: Why do we have to use MSBuild to call this target? I would expect that we are in the project and we should just call the FindBestConfiguration task directly.

File: buildvertical.targets [hunk 29]
```diff
@@ -8,102 +8,92 @@
...
   </Target>
 
   <!-- For the current project, target which determines compatibility for targets -->
   <Target Name="FindBestConfiguration"
-          Returns="@(VerticalTargets)"
+          Returns="@(_Configuration)"
           BeforeTargets="BeforeResolveReferences">
     
+    <Message Importance="Low" Text="FindBestConfiguration: $(MSBuildProjectFullPath) - $(BuildConfigurationImportFile)" />
     <MSBuild Targets="GetBuildConfigurations"
```
Comment (weshaggard) [inline buildvertical.targets]: Why do we need to call GetBuildConfigurations any longer here? We should just be able to get the BuildConfigurations property directly given we are in the project.

File: buildvertical.targets [hunk 30]
```diff
@@ -8,102 +8,92 @@
...
-      <Output TaskParameter="BestConfiguration" ItemName="VerticalTargets" />                          
+      <Output TaskParameter="BestConfiguration" ItemName="_Configuration" />                          
     </FindBestConfiguration>  
 
-    <ItemGroup Condition="'$(BuildConfigurations)' == ''">
-      <VerticalTargets Include="$(MSBuildProjectFullPath)">
-        <NoBuildConfigurations>true</NoBuildConfigurations>
-      </VerticalTargets>
+    <!-- We want to keep projects which do not have a BuildConfigurations specified and therefore do not return a valid Configuration -->
+    <ItemGroup Condition="'$(BuildConfigurations)' == '' and '@(_Configuration)' == ''">
```
Comment (weshaggard) [inline buildvertical.targets]: If BuildConfigurations is empty in what cases would _Configurations be set?

File: buildvertical.targets [hunk 31]
```diff
@@ -8,102 +8,92 @@
...
       <Output TaskParameter="TargetOutputs"
-              ItemName="VerticalTargets" />
+              ItemName="_Configuration" />
     </MSBuild>
 
     <PropertyGroup>
-      <_SelectedOSGroup>%(VerticalTargets.OSGroup)</_SelectedOSGroup>
-      <_SelectedTargetGroup>%(VerticalTargets.TargetGroup)</_SelectedTargetGroup>
-      <_KeepProject>%(VerticalTargets.NoBuildConfigurations)</_KeepProject>
+      <_BestConfiguration>%(_Configuration.Identity)</_BestConfiguration>
```
Comment (weshaggard) [inline buildvertical.targets]: This is making the assumption that we only have one configuration, which is OK for now but we will likely need address this later.

File: buildvertical.targets [hunk 32]
```diff
@@ -8,102 +8,92 @@
...
-      <Project Remove="@(OmitProject)" />
-    </ItemGroup>
+ <Target Name="AnnotateProjectsWithConfiguration" 
+         Returns="@(ProjectWithConfiguration)"> 
+    <PropertyGroup>
+      <_ProjectName>$(MSBuildProjectFullPath)</_ProjectName>
+    </PropertyGroup>
+    <Message Importance="Low" Text="$(MSBuildProjectFullPath), Configuration - $(_BestConfiguration)" />
+    <ItemGroup> 
+       <ProjectWithConfiguration Include="$(_ProjectName)" Exclude="@(OmitProject)">
```
Comment (weshaggard) [inline buildvertical.targets]: To support multiple configurations I think the best approach would be to batch over all of them here and add a ProjectWithConfiguration for each.

Comment (weshaggard) [inline buildvertical.targets]: The more I think about it the more I think you should just collapse all the independent targets you are calling down into just this one and take the direct output from FindBestConfiguration task and feed it to ProjectWithConfiguration here. That will eliminate a lot of the duplicated properties and assumptions across these targets as well.

File: buildvertical.targets (hunk 30 above)
Comment (chcosta) [inline buildvertical.targets]: it wouldn't, but...
<table>
  <tr>
    <td>Configuration</td>
    <td>BuildConfigurations</td>
    <td>Result</td>
  </tr>
  <tr>
    <td>something</td>
    <td>n/a</td>
    <td>valid</td>
  </tr>
  <tr>
    <td>empty</td>
    <td>something</td>
    <td>no valid configuration for vertical, remove from traversal</td>
  </tr>
  <tr>
    <td>empty</td>
    <td>empty</td>
    <td>valid project, with no configurations (retain in traversal)</td>
  </tr>
</table>
